File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.py
from copy import deepcopy

class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        outputs = []

        buf = []
        def search(buf, sm, idx):
            nonlocal outputs
            if sm > target or idx >= len(candidates): return

            buf.append(candidates[idx])
            if sm+candidates[idx] == target:
                outputs.append(deepcopy(buf))
            elif sm+candidates[idx] < target:
                search(buf, sm+candidates[idx], idx+1)
            
            buf.pop()
            while idx < len(candidates) - 1 and candidates[idx] == candidates[idx + 1]: idx += 1
            search(buf, sm, idx + 1)
        
        search(buf, 0, 0)
        return outputs
        

File: Data Structures & Algorithms/count-subsequences/submission-1.cpp
class Solution {
public:
    int numDistinct(string s, string t) {
        vector<int>dp(t.size());

        for (char c: s) {
            for (int i = t.size()-1; i >0; --i) {
                if (c == t[i]) {
                    dp[i] += dp[i-1];
                }
            }
            if (c == t[0]) ++dp[0];
        }
        return dp[t.size()-1];
    }
};
